### ISSUE TITLE
Tabs don't update when array order changes

### ISSUE THREAD

**Issue by pmorch** [bootstrap-vue/bootstrap-vue#3506]
### Describe the bug

`<b-tabs>` doesn't update reliably when the array changes.

* If the _number_ of elements changes, the update seems to work.
* If the _order_ of elements changes, it doesn't seem to update.

#1781 is (almost) a duplicate of this. In that issue, the user used a 

    <b-tab v-for="tab in tabs" :title="tab.title" :key="`key-${tab.id}`">

The bug was closed when @emanuelmutschlechner says:

> You must use the index of tabs array. By using the id as key, shuffling the tabs array will not change the underlying HTML at all.

As in:

    <b-tab v-for="(tab, index) in tabs" :title="tab.title" :key="`key-${index}`">

I think that reasoning is wrong. Forcing me to use index as key means that something is _broken_.

Superficially it seems to work. But as the Vue [documentation](https://vuejs.org/v2/guide/list.html#Maintaining-State)  points out:

> This default mode is efficient, but **only suitable when your list render output does not rely on child component state or temporary DOM state (e.g. form input values).**

(highlight theirs)

So if the different tabs have e.g. `<input>`s then Vue will get confused. That is exactly what `key` is for. If you use index as key, you might as well omit it altogether as using the index is the default behavior

The Vue documentation explicitly uses `item.id` in [several](https://vuejs.org/v2/guide/list.html#Maintaining-State) [places](https://vuejs.org/v2/api/#key). I don't find any `index` used for key anywhere.

And using `:key="item.id"`, `<b-tabs>` doesn't update when the sequence changes.

In my opinion, there is a bug here. `:key="item.id"` should work.

### Steps to reproduce the bug

See this [codepen](https://codepen.io/pmorch/pen/ZdGgJZ). It has inputs on the tabs and you can see that the inputs end up in the wrong tabs when the array order changes.

### Expected behavior

I expect:

    <b-tab v-for="tab in tabs" :title="tab.title" :key="`key-${tab.id}`">

to work whenever tabs changes - also if just the order changes.

### Versions

**Libraries:**

- BootstrapVue: 2.0.0-rc.22
- Bootstrap: 4.3.1
- Vue: 2.6.10

### Demo link

See this [codepen](https://codepen.io/pmorch/pen/ZdGgJZ)

**Comment by tmorehouse** [comment]
PR https://github.com/bootstrap-vue/bootstrap-vue/pull/3442 should have this fixed... which will be in release 2.0.0-rc.23 coming out soon.

**Comment by pmorch** [comment]
Thanks, I'll sit tight...

**Comment by tmorehouse** [comment]
But yeah, when removing a tab from an array, and using the array index as the key is not overly reliable as Vue doesn't realize the tabs have shifted indexes.

The PR uses the tab's Vue `_uid` for indexing/key, and DOM queries to ensure the detected child tabs  order matches DOM order.

**Comment by pmorch** [comment]
Uhm... What? 

Here is Evan You, creator of Vue, [saying](https://github.com/vuejs/vue/issues/5886#issuecomment-308625735):

> The vm _uid is reserved for internal use and it's important to keep it private (and not rely on it in user code) so that we keep the flexibility to change its behavior for potential future use cases.

To me it is now a little unclear: Once PR #3442 is released, will I be able to use _my own_ `key` as in:

    <b-tab v-for="tab in tabs" :title="tab.title" :key="`key-${tab.id}`">

?

If my _own_ `key` still is going to work, great. Otherwise, I'm curious (I haven't read the PR yet): How can BootstrapVue use the `_uid` to make bullet-proof key handling? Exactly what corner cases are going to fail? And if it is 100% bullet-proof so `key` isn't needed, why doesn't Vue itself (and [react too](https://reactjs.org/docs/lists-and-keys.html#keys)) use this new algorithm and deprecate the use of  `key` altogether?

**Comment by tmorehouse** [comment]
You can still use a key of you want. Or use unique ID assigned or use no key. The new detection routine relies on the rendered DOM for detecting the order of the tabs... and doesn't rely on $slots.default or $children to detect order (like it did in previous versions) uid is unique (even if how they implement it in the future... just it can't be relied upon between server and client)

So dont worry about it as the newer release you wont need to assign keys... un less you are doing some weird `:is` stuff or notice strange behaviour in your implementation of rendering tabs. Keys will help vue to stop re-using elements/vnodes. Just like in v-for loops...

**Comment by tmorehouse** [comment]
@pmorch v2.0.0-rc.23 has been released.

**Comment by pmorch** [comment]
This still doesn't work even with v2.0.0-rc.23 as shown in this updated ["BootstrapVue@2.0.0-rc.23 b-tabs don't follow data" codepen](https://codepen.io/pmorch/pen/LKpgPb). I therefore suggest to re-open this this issue.

Your comment from before actually has me quite a bit worried:

> So don't worry about it as the newer release you wont need to assign keys... unless you are doing some weird :is stuff or notice strange behaviour in your implementation of rendering/generating the b-tab components in your document. Keys will help Vue to stop re-using elements/vnodes. Just like in v-for loops...

I *am* doing something that to you is strange stuff (apparently) and encountering strange and buggy behavior... I still don't see how this project can avoid the user having to set up keys when neither Vue nor React have been able to avoid it. _And still have it work in every single case_. It worries me most that you cannot be 100% crisp and precise about what will work _and what won't_. Hand coding with Bootstrap 4 markup, Vue and `:key="tab.id"` gets it working 100% of the time.

As I asked before:

1. Do you agree that `key` is unavoidably **necessary** in some cases with `v-for` in vanilla Vue?
1. If you do, why isn't it also exactly likewise unavoidably **necessary** in some cases for `<b-tabs>`?

One test case to consider: Two completely identical tabs: Same markup, same contents everything, except for a different `tab.id`. They both contain an unbound `<input>` field, whose value isn't reflected in Vue at all. I manually enter different values in the input field in each tab.  If I re-order the array I want the inputs (containing temporary, non-Vue values) to be reordered also.

Also, when looking at `src/components/tabs/tabs.js` in the `BTabs` definition, I don't see `key` as a prop but I do see:

    const buttons = tabs.map((tab, index) => {
      <snip>
      return h(BTabButtonHelper, {
        key: tab._uid || index,
        <snip>

Am I missing something or is the `key` attribute completely ignored?

You write:

> You can still use a key of you want. Or uses the unique ID assigned (by user) to b-tab

Using an `ID` on a child element is not a viable substitute for the `key` attribute (since Vue could just try to change the ID attribute instead of correctly identify which tab matches which key.) Again, if it where, why does Vue and react even have the `key` attribute?

Background: We're using some tabs in Vue wrapping some legacy widgets written in vanilla jQuery javascript. When changing the tabs it is crucial to us that the underlying HTML and javascript in the widgets is left undisturbed and that Vue doesn't try to re-render them or monkey-patch the widget HTML. Which is possible and works with Vue when using `key`s. Here is a screenshot of the AngularJS 1.x / Bootstrap2 implementation that I'm trying to upgrade/refactor/rewrite:

![image](https://user-images.githubusercontent.com/159292/59493475-2c84a680-8e8b-11e9-9c50-6a2136c89ec0.png)

**Comment by tmorehouse** [comment]
`this.tabs = this.tabs.reverse()`

`Array.prototype.reverse` is an in-place method. I don't believe vue has a trap for that (like they do with other array mutation methods like splice, push, pop, shift, unshift).

Change to this:

`this.tabs = this.tabs.slice().reverse()` to create a new array reference.

**Comment by tmorehouse** [comment]
the `_uid` is used to generate IDs when no ID is provided

**Comment by emanuelmutschlechner** [comment]
@pmorch i fixed your codepen: https://codepen.io/anon/pen/EBPVWo

**Comment by pmorch** [comment]
Thank you for the time you've spent on this, @tmorehouse.

Why is it that you so haven't yet commented on my repeated questions:
1. Do you agree that `key` is unavoidably **necessary** in some cases with `v-for` in vanilla Vue?
1. If you do, why isn't it also exactly likewise unavoidably **necessary** in some cases for `<b-tabs>`?

You see, I don't think any local magic is possible short of real `key` support.

**Comment by emanuelmutschlechner** [comment]
@pmorch have a look at the official api docs for `:key` as it explains what it does: https://vuejs.org/v2/api/#key

**Comment by emanuelmutschlechner** [comment]
@pmorch if you're dealing with legacy html widgets I would recommend to wrap the legacy stuff for every tab in its own component. Your legacy component wrappers should word fine in combination with `keep-alive` (and keys).

https://vuejs.org/v2/guide/components-dynamic-async.html#keep-alive-with-Dynamic-Components

**Comment by tmorehouse** [comment]
@pmorch keys are required for list rendering to prevent unnecessary re-rending, and to ensure proper hild subcomponent state.  If you are creating your `b-tab` children with a `v-for` loop, then you most-likely need a key.  If you are placing your `b-tab` children manually inside a `b-tabs` component (even if using `v-if`/`v-else`) you don't need a key (as they are not created in a loop).

Keys may be needed if you are doing something similar v-for loops.  It is no longer required if **not** using a v-for loops.

Not all use cases for `b-tab` needs  keys is my point.  You don't have to provide them (except when using loops, but even then you may not need them if your tab content is simple and doesn't rely on states (i.e. textual content only, no internal state).

I am not saying you never ever need keys.  It just isn't needed all the time anymore (with `b-tabs`/`b-tab`). There were issues in earlier releases of bootstrap-vue where keys were indeed needed to get around some issues.  those issues have been fixed, which is why those issues are closed.

**Comment by pmorch** [comment]
Again, thank you for your time on this, @tmorehouse and @emanuelmutschlechner. I know something like this is not the favorite thing to spend time on for any of us.

To my surprise I have not been able to convince you that there is a bug here when

    <b-tab v-for="tab in tabs" :title="tab.title" :key="tab.id">

doesn't work as well as when I hard code Bootstrap markup with Vue and `:key="tab.id"`.

I'm sure you would also prefer that this issue died. I at least have spent _way_ too much time on this already (and have literally lost sleep over it) and I can't justify doing that any longer. I've become emotional about this and that serves nobody. So you've closed this issue. I've asked to have it re-opened, and you've declined. It is your project and your bugs. You decide.

I'll create the Bootstrap markup myself and use `v-for` with `:key="tab.id"` because that works where BootstrapVue doesn't and we can all be on our merry way.

If I learn something new, I'll update this issue or create a new one if I have something genuinely new to bring to the table.

**Comment by tmorehouse** [comment]
Have you tied the latest release 2.0.0-rc.23 which handles the case where tabs change order?

**Comment by pmorch** [comment]
As I wrote:

> This still doesn't work even with v2.0.0-rc.23 as shown in this updated ["BootstrapVue@2.0.0-rc.23 b-tabs don't follow data" codepen](https://codepen.io/pmorch/pen/LKpgPb). I therefore suggest to re-open this this issue.



That codepen still uses

    this.tabs = this.tabs.reverse()

Changing that to 

    this.tabs = this.tabs.slice().reverse()

Made no difference.

@emanuelmutschlechner "fixed" it to do this instead:

      Vue.$set(this, 'tabs', this.tabs.reverse());

And that looks like it works. However it spews out errors to the console. They say:

> "TypeError: Vue.$set is not a function"

And true enough, 

    this.tabs.reverse()

also works. So I think that is what @emanuelmutschlechner 's codepen actually ends up doing.

I still have no idea why the first two don't work when they work fine with vanilla Vue and Bootstrap using `key="tabs.id"`. And then I have no confidence in the solution. The thing is that I'm assuming that BootstrapVue is just a wrapper around Bootstrap in Vue and that the '<b-*>` components should work *exactly* as if I'd put in the much more verbose Bootstrap markup. When it doesn't, as in this issue, we have a leaky abstraction and I loose confidence in the whole thing.

I _actually_ want to make the tabs sortable/reoderable with VueDraggable and I have that working with vanilla Bootstrap markup (on the very first try) but it doesn't work with BootstrapVue. I'm presuming because of this issue.

**Comment by tmorehouse** [comment]
Since you are rendering with a v-for loop (from an array), you will need to provide a unique key for each tab (not based on index) so that Vue can re-use the correct vNodes, and since BootstrapVue does not use your array directly (nor rendering the loop for you), you are responsible for providing a unique `key`.

Aside: `Vue.set(...)` (no `$`) if using the Global Vue API, `this.$set(...)` if using the Vue instance version.

**Comment by tmorehouse** [comment]
You must also remember that Vue tried to minimize re-render of elements.  It patches existing elements (if possible) rather than trashing the existing elements and creating new.

https://vuejs.org/v2/guide/conditional.html#Controlling-Reusable-Elements-with-key

> Vue tries to render elements as efficiently as possible, often re-using them instead of rendering from scratch. Beyond helping make Vue very fast, this can have some useful advantages. ...
> ...
> This isn’t always desirable though, so Vue offers a way for you to say, “These two elements are completely separate - don’t re-use them.” Add a key attribute with unique values

**Comment by pmorch** [comment]
> Since you are rendering with a v-for loop (from an array, and altering their order), you will need to provide a unique key for each tab (not based on index) so that Vue can re-use the correct vNodes, and since BootstrapVue does not use your array directly (nor rendering the loop for you), you are responsible for providing a unique key.

My despair is now complete. Have you taken a look at my very first codepen in the OP, 20 comments ago? Or the  [BootstrapVue@2.0.0-rc.23](https://codepen.io/pmorch/pen/LKpgPb) version:

`tabs` in JSON:
```
[
  {
    "id": 1,
    "name": "tab1",
    "active": true
  },
  {
    "id": 2,
    "name": "tab2"
  }
]
```
template for the "B-tabs using tab.id - doesn't work" section:
```
  <b-tabs>
    <b-tab v-for="tab in tabs" :title="tab.name" :key="tab.id">
      {{tab.name}}
      <input :id="'btabs1-' + tab.id" type="text">
    </b-tab>
  </b-tabs>
```
Can we please once and for all agree that I *am* supplying a unique key? Is there _anything_ wrong with what I'm doing above? If I'm not providing a unique key, please say so directly and point to what I'm doing wrong. You've said now many times I can specify a key, but _I've been trying to do that all along_. If I wasn't supplying a unique key, then of course that would've been the problem.

I have no idea why you sent me the link to https://vuejs.org/v2/guide/conditional.html#Controlling-Reusable-Elements-with-key because as I read it, it proves _my_ point: input entered by the user in `<input>` markup under the same key values should remain unchanged and not have their values changed, just because the order of b-tabs changes. That is exactly the point of the key!

So to be clear and explicit: before I hit the "Reverse" button, the input on the b-tab with key=1 (from tabs[0]) has the value: "user enters 1 in tab 1". After I hit the "Reverse button" I expect the input on the b-tab with key=1 (now from tabs[1]) to still have the value: "user enters 1 in tab 1". Because of the above. Where in fact the input on the b-tab with key=1 has the value: "user enters 2 in tab 2" after the "Reverse". Hey, that "user enters 2 in tab 2" after the "Reverse" `<input>` should be in the b-tag with key=2 just like before I hit the button. Just like vanilla Bootstrap + Vue does.

My ***entire*** point is that the inputs in the b-tab for e.g. key=1 don't stay in the b-tab with key=1, but weirdly moves to the b-tab with key=2 and vice versa. 

That is the entire point and the core of this bug report!

**Comment by emanuelmutschlechner** [comment]
@pmorch i got the first `<b-tabs>` working the way you want them to: https://codepen.io/anon/pen/dBXpJE

just took a look at the [source](https://github.com/bootstrap-vue/bootstrap-vue/blob/63ef94e8ef8b5c102b2254074a0d066a441bc413/src/components/tabs/tabs.js#L404) implementation of tabs and calling `updateTabs` the next frame does exactly what you want. it forces the component to update its children.

**Comment by tmorehouse** [comment]
My mistake... it is the upcoming 2.0.0-rc.24 that has the fix, not the current 2.0.0-rc.23

https://github.com/bootstrap-vue/bootstrap-vue/pull/3529

**Comment by pmorch** [comment]
@emanuelmutschlechner: I can see that your codepen now does:

    this.$refs.btabs1.updateTabs());

And yes, it seems to work. I thank you for (again) putting in the effort in helping me getting something that works. Seriously, thank you.

Wouldn't  you agree that something like the above smells like working around a bug?

Code littered with this kind of thing is brittle. Otherwise a challenge for you: Please write the comment that goes above that line explaining why it is required - other than "It seems to do the trick with the current version of Bootstrap.Vue". I definitely expect it to work without such manual update code.

And at this point I'm not only looking for code that works. I'm looking for code that I understand why works and that I can explain to others. I already have both those with vanilla Bootstrap and Vue. But want to use new and shiny Bootstrap.Vue! :-)

**Comment by tmorehouse** [comment]
2.0.0-rc.24 has been released

**Comment by emanuelmutschlechner** [comment]
@pmorch check the updated codepen [BootstrapVue@2.0.0-rc.24](https://codepen.io/pmorch/pen/LKpgPb)

the new RC24 does indeed fix this bug

**Comment by pmorch** [comment]
@tmorehouse: I just built dev from the current git repo and hosted it myself somewhere and it seems to work now. What made it work was commit 130f8ffe85305a4f6eb9f38e199562edeabf72ea.

I now get it.

`<b-tabs>` is letting Vue do the ordering of the `<b-tab>` instances that contain the tab _content_ in the default slot. And they're ordered correctly because of the `key` on `<b-tab>`. But BootstrapVue isn't involved with any of that. In tabs.js, this.getTabs() aligns the order of the buttons with DOM order of tabs previously set up by Vue in the default slot.

What I was struggling with was that searching for `key` in `src/components/tabs/*` turned up so precious little, as I was assuming (erroneously), that somehow the key attribute needed to be passed on somewhere.  So I was thinking "the key attribute I supply isn't consumed/handled anywhere". But no. default slot and Vue takes care of all that perfectly. Once that is done, BTabs/tabs.js just mirrors the DOM order in getTabs().

The fact that the order of buttons was wrong led me to believe that the `key` attribute "wasn't working". What I now understand is that the `key` attribute the entire time correctly determined the order inside `.tab-content`, but that the `ul.nav-tabs` didn't reliably mirror that (correct) DOM ordering.

And so now with 130f8ffe85305a4f6eb9f38e199562edeabf72ea it (looks like) it observes/watches `this.$refs.tabsContainer` and calls `this.updateTabs.bind(this)` when that changes. (Essentially the same fix as @emanuelmutschlechner put in manually.)

My faith in the universe is restored.

I hope I have not done irreparable damage here. I apologize for being too stubborn and just generally a pain the behind. I agree it looks like this will be fixed in 2.0.0-rc.24.

**Comment by tmorehouse** [comment]
BootstrapVue 2.0.0-rc.24 has been published about an hour ago: https://www.npmjs.com/package/bootstrap-vue
